server/cluster/cluster.go

```diff
@@ -1032,43 +1072,72 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo, tracer cor
-		if overlaps, err = c.core.AtomicCheckAndPutRegion(region, tracer); err != nil {
+		if overlaps, err = c.core.AtomicCheckAndPutRegion(ctx, region); err != nil {
 			tracer.OnSaveCacheFinished()
 			return err
 		}
 		if !c.IsServiceIndependent(mcsutils.SchedulingServiceName) {
-			cluster.HandleOverlaps(c, overlaps)
+			runner.RunTask(
+				ctx,
+				ratelimit.TaskOpts{
+					TaskName: "HandleOverlaps",
+					Limit:    limiter,
+				},
+				func(ctx context.Context) {
+					cluster.HandleOverlaps(c, overlaps)
+				},
+			)
 		}
 		regionUpdateCacheEventCounter.Inc()
 	}
 
 	tracer.OnSaveCacheFinished()
-	// TODO: Due to the accuracy requirements of the API "/regions/check/xxx",
-	// region stats needs to be collected in API mode.
-	// We need to think of a better way to reduce this part of the cost in the future.
-	cluster.Collect(c, region, c.GetRegionStores(region), hasRegionStats)
+	// handle region stats
+	runner.RunTask(
+		ctx,
+		ratelimit.TaskOpts{
+			TaskName: "CollectRegionStatsAsync",
+			Limit:    c.hbConcurrencyLimiter,
+		},
+		func(ctx context.Context) {
+			// TODO: Due to the accuracy requirements of the API "/regions/check/xxx",
+			// region stats needs to be collected in API mode.
+			// We need to think of a better way to reduce this part of the cost in the future.
+			cluster.Collect(c, region, hasRegionStats)
+		},
+	)
+
 	tracer.OnCollectRegionStatsFinished()
 	if c.storage != nil {
-		// If there are concurrent heartbeats from the same region, the last write will win even if
-		// writes to storage in the critical area. So don't use mutex to protect it.
-		// Not successfully saved to storage is not fatal, it only leads to longer warm-up
-		// after restart. Here we only log the error then go on updating cache.
-		for _, item := range overlaps {
-			if err := c.storage.DeleteRegion(item.GetMeta()); err != nil {
-				log.Error("failed to delete region from storage",
-					zap.Uint64("region-id", item.GetID()),
-					logutil.ZapRedactStringer("region-meta", core.RegionToHexMeta(item.GetMeta())),
-					errs.ZapError(err))
-			}
-		}
 		if saveKV {
-			if err := c.storage.SaveRegion(region.GetMeta()); err != nil {
-				log.Error("failed to save region to storage",
-					zap.Uint64("region-id", region.GetID()),
-					logutil.ZapRedactStringer("region-meta", core.RegionToHexMeta(region.GetMeta())),
-					errs.ZapError(err))
-			}
-			regionUpdateKVEventCounter.Inc()
+			runner.RunTask(
+				ctx,
+				ratelimit.TaskOpts{
+					TaskName: "SaveRegionToKV",
+					Limit:    c.hbConcurrencyLimiter,
+				},
+				func(ctx context.Context) {
+					// If there are concurrent heartbeats from the same region, the last write will win even if
+					// writes to storage in the critical area. So don't use mutex to protect it.
+					// Not successfully saved to storage is not fatal, it only leads to longer warm-up
+					// after restart. Here we only log the error then go on updating cache.
+					for _, item := range overlaps {
```

Comment:
why move `for range` info `saveKV`?

Reply:
Combine it into a single task to execute simultaneously, which also helps reduce issues with parallel execution order.